File: core/src/main/java/edu/knowitall/taggers/tag/Tagger.java
package edu.knowitall.taggers.tag;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.jdom2.Element;

import edu.knowitall.collection.immutable.Interval;
import edu.knowitall.taggers.XmlSerializable;
import edu.knowitall.taggers.Type;
import edu.knowitall.taggers.constraint.*;
import edu.knowitall.tool.chunk.ChunkedToken;
import edu.knowitall.tool.stem.Lemmatized;

/***
 * A class that searches over sentences and tags matching text with a
 * type.
 * @author schmmd
 *
 */
public abstract class Tagger implements XmlSerializable {
    public final /*interned*/ String descriptor;
    public final /*interned*/ String source;
    protected List<Constraint> constraints;

    public Tagger(String descriptor, String source) {
        this(descriptor, source, new ArrayList<Constraint>());
    }

    public Tagger(String descriptor, String source, List<Constraint> constraints) {
        this.descriptor = descriptor.intern();
        this.source = source == null ? null : source.intern();
        this.constraints = constraints;
    }

    public abstract void sort();

    @Override
    public boolean equals(Object that) {
        if (that == null) return false;
        if (this == that) return true;
        if (this.getClass() != that.getClass()) return false;

        Tagger tagger = (Tagger)that;
        return this.descriptor.equals(tagger.descriptor);
    }

    /***
     * Public method for finding tags in a sentence.
     * @param sentence
     * @return a list of the tags found
     */
    public List<Type> tags(List<Lemmatized<ChunkedToken>> sentence) {
        return tags(sentence, Collections.<Type>emptyList());
    }

    /***
     * Public method for finding tags in a sentence with types.
     * This method also filters out types by constraint.
     *
     * @param sentence
     * @return a list of the tags found
     */
    public List<Type> tags(List<Lemmatized<ChunkedToken>> sentence, List<Type> types) {
        List<Type> tags = findTagsWithTypes(sentence, types);

        // remove types that are covered by other types.
        filterCovered(tags);
        tags = filterWithConstraints(sentence, tags);

        return tags;
    }

    protected abstract List<Type> findTags(List<Lemmatized<ChunkedToken>> sentence);

    /**
     * This method should be overridden by any Tagger that wants to use the
     * Types accumulated from previous Taggers. If it's not overridden the sentence
     * will be tagged without type information.
     * @param sentence
     * @param types
     * @return
     */
    public List<Type> findTagsWithTypes(List<Lemmatized<ChunkedToken>> sentence, List<Type> types){
        return findTags(sentence);
    }

    /***
     * Remove types that cover over types.
     * @param tags
     */
    public void filterCovered(List<Type> tags) {
        for (int i = 0; i < tags.size(); i++) {
            for (int j = 0; j < tags.size(); j++) {
                if (i != j) {
                    Type tagi = tags.get(i);
                    Type tagj = tags.get(j);
                    if (tagi.descriptor().equals(tagj.descriptor()) &&
                        tagi.interval().superset(tagj.interval())) {
                        tags.set(j, tags.get(tags.size() - 1));
                        tags.remove(tags.size() - 1);

                        if (i >= tags.size()) {
                            break;
                        }
                    }
                }
            }
        }
    }

    public List<Type> filterWithConstraints(List<Lemmatized<ChunkedToken>> sentence, List<Type> tags) {
        List<Type> filtered = new ArrayList<Type>();
        for (Type tag : tags) {
            boolean passesConstraints = true;
            for (Constraint constraint : this.constraints) {
                if (!constraint.apply(sentence.subList(tag.interval().start(), tag.interval().end()), tag)) {
                    passesConstraints = false;
                    break;
                }
            }

            if (passesConstraints) {
                filtered.add(tag);
            }
        }

        return filtered;
    }

    public Type createType(List<Lemmatized<ChunkedToken>> sentence, Interval interval) {
        return Type.fromSentence(sentence, this.descriptor, this.source, interval);
    }

    public void constrain(Constraint constraint) {
        this.constraints.add(constraint);
    }

    public static Class<?> getTaggerClass(String classname, String pack) throws ClassNotFoundException {
        try {
          return Class.forName(pack + "." + classname.replace('.', '$'));
        }
        catch(ClassNotFoundException ex) {
          return Class.forName(classname.replace('.', '$'));
        }
    }

    public static Tagger create(String classname, String pack, Object[] argValues) throws ClassNotFoundException, SecurityException, NoSuchMethodException, IllegalArgumentException, InstantiationException, IllegalAccessException, InvocationTargetException {
        return create(getTaggerClass(classname, pack), new Class<?>[] { String.class, List.class }, argValues);
    }

    public static Tagger create(String classname, String pack, Class<?>[] argTypes, Object[] argValues) throws ClassNotFoundException, SecurityException, NoSuchMethodException, IllegalArgumentException, InstantiationException, IllegalAccessException, InvocationTargetException {
        return create(getTaggerClass(classname, pack), argTypes, argValues);
    }

    public static Tagger create(Class<?> tagger, Class<?>[] argTypes, Object[] argValues) throws ClassNotFoundException, SecurityException, NoSuchMethodException, IllegalArgumentException, InstantiationException, IllegalAccessException, InvocationTargetException {
        Constructor<?> constructor = tagger.getConstructor(argTypes);
        return (Tagger)constructor.newInstance(argValues);
    }

    public static Tagger create(Element element) throws SecurityException, IllegalArgumentException, ClassNotFoundException, NoSuchMethodException, InstantiationException, IllegalAccessException, InvocationTargetException {
        String classname = element.getName();

        String pack = element.getAttributeValue("package");
        if (pack == null) {
            pack = Tagger.class.getPackage().getName();
        }

        return Tagger.create(classname, pack, new Class[] { Element.class }, new Object[] { element });
    }

    @Override
    public String toString() {
        return this.getClass().getSimpleName() + "(" + this.descriptor + ")";
    }

    /// XML

    /***
     * Deserialize from an XML element.
     */
    public Tagger(Element e) throws ParseTagException {
        this.descriptor = e.getAttributeValue("descriptor");
        this.source = e.getAttributeValue("source");
        if (this.descriptor == null) {
            throw new ParseTagException("No attribute 'descriptor'", e);
        }

        List<Constraint> constraints = new ArrayList<Constraint>();
        List<Element> constraintElements = e.getChildren("constraint");
        for (Element element : constraintElements) {
            String type = element.getAttributeValue("type");
            constraints.add(Constraint.create(type));
        }

        this.constraints = constraints;
    }

    /***
     * Serialize to an XML element.
     */
    @Override
    public Element toXmlElement() {
        String nameWithoutPackage = this.getClass().getName().substring(this.getClass().getPackage().getName().length() + 1).replace('$', '.');
        Element e = new Element(nameWithoutPackage);

        if (!this.getClass().getPackage().getName().equals(Tagger.class.getPackage().getName())) {
            e.setAttribute("package", this.getClass().getPackage().getName());
        }

        e.setAttribute("descriptor", this.descriptor);

        for (Constraint constraint : constraints) {
            e.addContent(new Element("constraint").setAttribute("type", constraint.getClass().getSimpleName()));
        }

        return e;
    }
}

File: core/src/main/scala/edu/knowitall/taggers/tag/Tagger.scala
package edu.knowitall.taggers.tag

import edu.knowitall.common.HashCodeHelper
import edu.knowitall.repr.sentence.Sentence
import edu.knowitall.taggers.constraint.Constraint
import edu.knowitall.tool.chunk.ChunkedToken
import edu.knowitall.tool.stem.Lemmatized
import edu.knowitall.tool.typer.Type
import edu.knowitall.tool.typer.Typer

import scala.collection.JavaConverters._

/** A tagger operations on a sentence to create types. */
abstract class Tagger[-S <: Sentence] {
  type TheSentence = S

  def name: String
  def source: String

  def sort(): this.type = this

  override def toString = this.getClass.getName + ":=" + name
  override def equals(that: Any) = that match {
    // fast comparison for Intervals
    case that: Tagger[_] => that.canEqual(this) && that.name == this.name && that.source == this.source
    // slower comparison for Seqs
    case that: IndexedSeq[_] => super.equals(that)
    case _ => false
  }
  def canEqual(that: Any) = that.isInstanceOf[Tagger[_]]

  override def hashCode = HashCodeHelper(name /*, constraints*/ )

  /** Public method for finding tags in a sentence.
    * @param sentence
    * @return a list of the tags found
    */
  def apply(sentence: S): Seq[Type] = {
    this.apply(sentence, Seq.empty, Seq.empty)
  }

  /** Public method for finding tags in a sentence with types.
    * This method also filters out types using the specified constraints.
    *
    * @param sentence
    * @param types  types already existing in the sentence
    * @param consumedIndices  indices used up on a previous level in a cascade
    * @return a list of the tags found
    */
  def apply(sentence: S, types: Seq[Type], consumedIndices: Seq[Int]): Seq[Type] = {
    var tags = findTagsWithTypes(sentence, types, consumedIndices)

    // remove types that are covered by other types.
    tags = filterCovered(tags)
    tags = filterWithConstraints(sentence, tags)

    tags
  }

  private[tag] def findTags(sentence: S): Seq[Type]

  // TODO(schmmd): one shouldn't need to override a method to provide a correct implementation
  /** This method should be overridden by any Tagger that wants to use the
    * Types accumulated from previous Taggers. If it's not overridden the sentence
    * will be tagged without type information.
    *
    * @param sentence
    * @param types  types already existing in the sentence
    * @param consumedIndices  indices used up on a previous level in a cascade
    *
    * @return
    */
  private[tag] def findTagsWithTypes(sentence: S, types: Seq[Type], consumedIndices: Seq[Int]): Seq[Type] = {
    findTags(sentence)
  }

  /** Remove types that cover over types.
    * @param tags
    */
  private def filterCovered(tags: Seq[Type]): Seq[Type] = {
    tags.filter { tag =>
      tags.find { other =>
        other != tag &&
          other.name == tag.name &&
          other.source == tag.source &&
          (other.tokenInterval superset tag.tokenInterval)
      } match {
        case Some(superType) => false
        case None => true
      }
    }
  }

  /** Remove types that do not pass the constraints.
    */
  private def filterWithConstraints(sentence: S, types: Seq[Type]) = {
    for {
      tag <- types
      // if this.constraints.forall(_(sentence, tag))
    } yield (tag)
  }
}

object Tagger {
  def getTaggerClass(classname: String, pack: String): Class[_] = {
    try {
      Class.forName(pack + "." + classname.replace('.', '$'))
    } catch {
      case ex: ClassNotFoundException =>
        Class.forName(classname.replace('.', '$'))
    }
  }

  def create[S <: Sentence](classname: String, pack: String, name: String, args: Seq[String]): Tagger[S] = {
    create[S](getTaggerClass(classname, pack), Array[Class[_]](classOf[String], classOf[Seq[String]]), Array[Object](name, args))
  }

  def create[S <: Sentence](classname: String, pack: String, argTypes: Array[Class[_]], argValues: Array[Object]): Tagger[S] = {
    create[S](getTaggerClass(classname, pack), argTypes, argValues)
  }

  def create[S <: Sentence](tagger: Class[_], argTypes: Array[Class[_]], argValues: Array[Object]): Tagger[S] = {
    val constructor = tagger.getConstructor(argTypes: _*)
    constructor.newInstance(argValues: _*).asInstanceOf[Tagger[S]]
  }
}
